torch/nn/parallel/distributed.py

```diff
@@ -88,6 +89,204 @@ def _dump_DDP_relevant_env_vars():
+            # process.
+            zero_tensors = [
+                torch.zeros_like(t).to(
+                    self.net.device_ids[0] if self.net.device_type != "cpu" else "cpu"
+                )
+                for t in bucket_tensors
+            ]
+            work = self.pg.allreduce(zero_tensors, AllreduceOptions())
+            allreduce_work.append(work)
+        for work in allreduce_work:
+            work.wait()
+
+    # Allreduces the used parameter mapping across ranks.
+    def _match_unused_params_allreduce(self):
+        locally_used_param_maps = self.net.reducer._get_local_used_maps()
+        self.pg.allreduce(locally_used_param_maps, AllreduceOptions())
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # Handle errors during training by propagating them.
+        if exc_val is not None:
+            raise exc_val
+
+        if not self.enable:
+            return
+        all_procs_joined = False
+        is_last_joiner = True
+        buckets_rebuilt = False
+        # Schedule an allreduce to match fwd pass allreduce in non-joined procs
+        while not all_procs_joined:
+            num_active_procs = self._schedule_all_reduce_for_fwd_pass()
+            if num_active_procs == 0:
+                all_procs_joined = True
+            else:
+                # Some DDP process still needs to be joined.
+                if is_last_joiner:
+                    is_last_joiner = False
+                # Schedule a corresponding broadcast if we are syncing module
+                # buffers in the forward pass.
+                self._check_and_sync_module_buffers()
+                # Schedule an allreduce to match the backwards pass allreduce.
+                self._match_all_reduce_for_bwd_pass()
+                # Check if we need to allreduce locally unused params.
+                if self.net.find_unused_parameters:
+                    self._match_unused_params_allreduce()
+                # Check if we need to rebuild buckets to match broadcast calls
+                # from other processes.
+                if not buckets_rebuilt:
+                    if self.net.reducer._should_rebuild_buckets():
+                        if self.net.find_unused_parameters:
+                            raise ValueError(
+                                "Rebuilding buckets with find_unused_parameters=True is not supported."
+                            )
+                        self.net.reducer._push_all_rebuilt_params()
+                        # Will rebuild and re-intialize tensor/bucket mapping so
+                        # that future allreduce orders are consisted across
+                        # joined and un-joined processes.
+                        self.net.reducer._rebuild_buckets()
+                    # Buckets were either just rebuilt or already rebuilt before
+                    # this process joined.
+                    buckets_rebuilt = True
```

Comment:
I'm wondering if we should have just a single API called `rebuildBuckets` which in C++ rebuilds the buckets if necessary. That way the reducer and the code here can just call the same API instead of having similar logic here.

Reply:
Good point, this will make the code a lot cleaner. Although, it may be good to keep a check here if the buckets are rebuilt or not. The reason is, having to call this from Python is a pretty special edge case where a rank terminates without iterating at all (if we iterate >= 1 times then we're guaranteed to have a call to rebuild buckets). Keeping this check will reduce the calls into C++ checking and possibly rebuilding buckets.